IBM/carbon-addons-iot-react
# feat(BarChartCard): variables, units, isExpanded, isEditable

Closes #1248 

**Summary**

Adds full support that TimeSeriesCard currently has:
  - Dynamic cardVariables support
  - Units to be shown in the legend and tooltip
  -  Tooltip that shows the dataset name, value, unit, color of dataset, and time
  - `isExpanded` state with StatefulTable displaying the chart data
    - For time-based charts, the data will be sorted by timestamp
    - For group charts, the data will be sorted by the grouped category
    - For simple charts, it will be a single row with the defined data
  - `isEditable` state with sample / preview data in grayscale colors (similar to `TimeSeriesCard`)
  - Adds explicit default colors if the user doesn't define them (the same ones carbon charts used)

**Change List (commits, features, bugs, etc)**

- Moved all bar chart utility functions to `barChartUtils.js`
- Added and updated tests / stories to reflect new features
- Moved `valueFormatter` from `timeSeriesUtils` to common `cardUtils`
- Added a check for null values before formatting the data as it will cause the graph to blow up
- Added stricter card sizes as anything below MEDIUM is unreadable
- Added BarChartCard to CardRenderer and Dashboard
- Added custom propType warnings for unique cases that will cause unexpected side effects
- Added default colors to lines if not explicitly defined
- Added color band to tooltip similar to `TimeSeriesCard`

**Acceptance Test (how to verify the PR)**

1. Go to all BarChartCard stories and see that they still function normally
  - Try the `isEditable` and `isExpandable` knobs on each story to see them working
2. Go to Dashboard stories and see that BarChartCard is functioning and displaying correctly

**Screenshots**

`isEditable`

![Screen Shot 2020-06-05 at 1 59 00 PM](https://user-images.githubusercontent.com/25177649/83913278-e5935f00-a734-11ea-9729-5c705146ba60.png)

`isExpanded`

![Screen Shot 2020-06-05 at 1 59 57 PM](https://user-images.githubusercontent.com/25177649/83913297-edeb9a00-a734-11ea-9390-8eff01ae0131.png)

Regular tooltip

![Screen Shot 2020-06-05 at 2 04 39 PM](https://user-images.githubusercontent.com/25177649/83913778-9863bd00-a735-11ea-9788-84edf9ba6365.png)

Time series tooltip

![Screen Shot 2020-06-05 at 2 04 51 PM](https://user-images.githubusercontent.com/25177649/83913789-9e599e00-a735-11ea-8c58-e6675fe6fc86.png)
